src/dmd/expressionsem.d

```diff
@@ -6084,7 +6084,17 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
             printf("AssertExp::semantic('%s')\n", exp.toChars());
         }
 
-        const generateMsg = !exp.msg && global.params.checkAction == CHECKACTION.context && global.params.useAssert == CHECKENABLE.on;
+        TemplateInstance ti;
+        const generateMsg = !exp.msg
+            && global.params.checkAction == CHECKACTION.context
+            // Don't generated a message if the assert is discarded
+            && global.params.useAssert == CHECKENABLE.on
+            // Only create instances in root modules or templates instantiated from those
+            // seperately compiled modules may use other -checkaction configurations, i.e.
+            // -checkaction=[halt|C|D] / -release / -check=assert=off
+            // => Missing template instances expected from their object file
+            && ((ti = sc.func.isInstantiated()) !is null ? ti.toParentLocal() : sc.func).getModule().isRoot();
```

Comment:
I think a working `-allinst` would take care of `-checkaction=context` problems. The price might be a bit high though, so disabling codegen culling of `_d_assert_fail` instantiations would be another hacky option.

Reply:
> But AFAIK, this won't work - the compiler relies on instantiations with equivalent params to behave identically, regardless of where they are instantiated. 

Right, this won't work (== produce unexpected results) for multiple instantiations in root and non-root modules.

> I think a working `-allinst` would take care of `-checkaction=context` problems. The price might be a bit high though, so disabling codegen culling of `_d_assert_fail` instantiations (incl. nested instantiations) would be another hacky option.

Thought about that too, but `-checkaction=context` should work without additionally passing `-allinst` - it incurs unnecessary overhead for larger projects (and not to mention the possibility of additional linker errors).

I've added another commit that tries to emit all instances from template declarations found in `core.internal.dassert` (i.e. a local `-allinst`). The patch doesn't work yet because several helper templates (e.g. `miniFormat`) aren't consistently added to the `members` of `core.internal.dassert`.

Reply:
> The patch doesn't work yet because several helper templates (e.g. miniFormat) aren't consistently added to the members of core.internal.dassert.

No template instances are ever appended to some parent instance's members - they are all appended to one module, no hierarchy. The only connection to the parent instance AFAIK is the `tinst` parent, which is checked in `needsCodegen`: https://github.com/dlang/dmd/blob/b6985bb2896477033bd34760dc7b9665c315c576/src/dmd/dtemplate.d#L6297-L6308

So by returning true in needsCodegen for *any* `_d_assert_fail` instantiation with `-checkaction=context` should probably suffice, for its children too.